core/db/db.go

```diff
@@ -52,59 +60,25 @@ type OpenchainDB struct {
 	StateDeltaCF *gorocksdb.ColumnFamilyHandle
 	IndexesCF    *gorocksdb.ColumnFamilyHandle
 	PersistCF    *gorocksdb.ColumnFamilyHandle
+	dbState      dbState
+	mux          sync.Mutex
 }
 
 var openchainDB *OpenchainDB
-var isOpen bool
+var once sync.Once
 
-// CreateDB creates a rocks db database
-func CreateDB() error {
-	dbPath := getDBPath()
-	dbLogger.Debug("Creating DB at [%s]", dbPath)
-	missing, err := dirMissingOrEmpty(dbPath)
-	if err != nil {
-		return err
-	}
-
-	if !missing {
-		return fmt.Errorf("db dir [%s] already exists", dbPath)
-	}
-	err = os.MkdirAll(path.Dir(dbPath), 0755)
-	if err != nil {
-		dbLogger.Error("Error calling  os.MkdirAll for directory path [%s]: %s", dbPath, err)
-		return fmt.Errorf("Error making directory path [%s]: %s", dbPath, err)
-	}
-	opts := gorocksdb.NewDefaultOptions()
-	defer opts.Destroy()
-	opts.SetCreateIfMissing(true)
-
-	db, err := gorocksdb.OpenDb(opts, dbPath)
-	if err != nil {
-		return err
-	}
-
-	defer db.Close()
-
-	dbLogger.Debug("DB created at [%s]", dbPath)
-	return nil
+// Create create an openchainDB instance
+func Create() *OpenchainDB {
+	return &OpenchainDB{dbState: closed}
 }
 
-// GetDBHandle returns a handle to OpenchainDB
+// GetDBHandle get an opened openchainDB singleton
 func GetDBHandle() *OpenchainDB {
-	var err error
-	if isOpen {
-		return openchainDB
-	}
-
-	err = createDBIfDBPathEmpty()
-	if err != nil {
-		panic(fmt.Sprintf("Error while trying to create DB: %s", err))
-	}
+	once.Do(func() {
+		openchainDB = Create()
+	})
+	openchainDB.Open()
```

Comment:
- A call to `GetDBHandle()`would acquire two locks. Can you make it one instead. `Create` alone is not of any use because, we never close and then open db anywhere.
- Can you check where else in the code, this method has been used and the db handle can be  cached. Till now, this method was not synchronized, hence caching was not meaningful.

@binhn this is yet another example where in fact this singleton can be created in `init` function and then would not require any locks. (Previously, we saw the similar example where it would have been better to reading configurations via `Viper` and cache in the `init`). However, the issue is that `init(s)` are invoked before `main` (and in the current code, `main` function loads the yaml file). Would it make sense to have a separate module for config and in the init of that module load yaml. This module would need wrapper apis that other modules would use instead of viper apis for getting configurations (in order to make other modules dependent on this configuration loading module so that it's init will be invoked before others.
The benefit would be that we would not require locking for singletons but the downside is having a wrapper over viper APIs.